src/com/ms/silverking/cloud/meta/ServerSet.java

```diff
@@ -18,112 +18,111 @@
-        return parse(new FileInputStream(file), VersionedDefinition.NO_VERSION);
-    }
-    
-    public static ServerSet parse(File file, long version) throws IOException {
-        return parse(new FileInputStream(file), version);
-    }
-    
-    public static ServerSet parse(InputStream in, long version) throws IOException {
-        return new ServerSet(ImmutableSet.copyOf(filterExtraText(StreamParser.parseLines(in))), 
-                                 version);
-    }
-    
-    private static List<String> filterExtraText(List<String> defs) {
-        List<String>    filtered;
-        
-        filtered = new ArrayList<>(defs.size());
-        for (String def : defs) {
-            String[]    d;
-            
-            d = def.split("\\s+");
-            filtered.add(d[0]);
-        }
-        return filtered;
-    }
-    
-    @Override
-    public boolean equals(Object o) {
-        return servers.equals(((ServerSet)o).servers);
+  private static Set<String> createCleanCopy(Set<String> servers) {
+    ImmutableSet.Builder<String> _servers;
+
+    _servers = ImmutableSet.builder();
+    for (String server : servers) {
+      String cleanDef;
+
+      cleanDef = cleanServerDef(server);
+      if (cleanDef.length() > 0) {
+        _servers.add(server);
+      }
     }
-    
-    @Override
-    public String toString() {
-        return StringUtil.toString(servers, '\n');
+    return _servers.build();
+  }
+
+  private static String cleanServerDef(String server) {
+    return server.toLowerCase().trim();
+  }
+
+  public int size() {
+    return servers.size();
+  }
+
+  public Set<String> getServers() {
+    return servers;
+  }
+
+  @Override
+  public long getVersion() {
+    return version;
+  }
+
+  public ServerSet add(Set<String> newServers) {
+    ImmutableSet.Builder<String> builder;
+
+    builder = new ImmutableSet.Builder<>();
+    ServerSet ss = new ServerSet(builder.addAll(servers).addAll(newServers).build(), getVersion());
+    return ss;
+    //return new ServerSet(builder.addAll(servers).addAll(newServers).build(), getVersion());
+  }
+
+  public ServerSet remove(Set<String> removedServers) {
+    Set<String> _s;
+
+    _s = new HashSet<>(servers);
+    _s.removeAll(removedServers);
+    return new ServerSet(ImmutableSet.copyOf(_s), getVersion());
+  }
+
+  public boolean contains(String id) {
+    return servers.contains(id);
+  }
+
+  public List<Node> filter(List<Node> raw) {
+    List<Node> filtered;
+
+    filtered = new ArrayList<>(raw.size());
+    for (Node node : raw) {
+      if (!servers.contains(node.getIDString())) {
+        filtered.add(node);
+      }
+    }
+    return filtered;
+  }
+
+  public static ServerSet parse(File file) throws IOException {
+    return parse(new FileInputStream(file), VersionedDefinition.NO_VERSION);
+  }
+
+  public static ServerSet parse(File file, long version) throws IOException {
+    return parse(new FileInputStream(file), version);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when ServerSet.parse throws an exception. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).